rust/agents/relayer/src/merkle_tree_builder.rs

```diff
@@ -8,20 +8,63 @@ use color_eyre::eyre::Result;
 use ethers::core::types::H256;
 use std::fmt::Display;
 
-use tracing::{debug, error};
+use tracing::{debug, error, info};
+
+// Helper type to denote the fact that a Inbox that has received no checkpoint is different from an Inbox that has been checkpointed with checkpoint index 0
+#[derive(Debug, Copy, Clone)]
+pub struct OnchainCheckpointIndex {
+    index: Option<u32>,
+}
+
+impl OnchainCheckpointIndex {
+    pub fn from_checkpoint(checkpoint: &Checkpoint) -> Self {
```

Comment:
Any thoughts on `impl From<&Checkpoint> for OnchainCheckpointIndex` instead?

And maybe `impl From<u32> for OnChainCheckpointIndex` instead of `from_index` down below?